Added Odd even category, changed dataset addition technique

RudraNilBasu opened the pull request:
* Added an odd-even category in the categorization activity

* changed the techniques used to add datasets in the activity. 
Originally the datasets were named category<xx>.qml. Now, the 
datasets can have any name, the FileNames::getFiles(String)
returns the list of files in the location, and using that, the datasets
are extracted, and using the isEmbedded variable, we decided 
whether the dataset is a demo or not.

petitlapin commented on src/activities/categorization/Categorization.qml:
```diff
@@ -36,6 +36,7 @@ ActivityBase {
 
     property string boardsUrl: "qrc:/gcompris/src/activities/categorization/resource/"
     property bool vert: background.width < background.height
+    property string datasetURL: "../../GCompris-qt/src/activities/categorization/resource/board/"
```
you need to use the qrc path, not a relative path, like boardsUrl

RudraNilBasu replied:
The qrc path was not working with opendir(). It returned NULL on using the qrc path.

petitlapin replied:
Using opendir is a bad choice here: it's not portable and it won't work once the software is installed. ../GCompris-qt works for you because the source code of gcompris is on this path/ Once installed, the source code is not installed so you no more have access to this path.

RudraNilBasu replied:
I shifted from using opendir() to QDir, but it still doesn't work for the qrc path.
Code - http://pastebin.com/JavNCD4G

RudraNilBasu replied:
Is there anything I am missing here ?

petitlapin commented on src/activities/categorization/categorization.js:
```diff
@@ -51,11 +51,18 @@ function init(items_,boardsUrl_) {
     items.categoryReview.stop()
+
+    var s = items.categories.split(" ")
+    var flag = items.file.exists(fileName) ? 1 : 0
```
if it is a boolean, use true and false. And a real name for the variable

petitlapin commented on src/activities/categorization/categorization.js:
```diff
@@ -51,11 +51,18 @@ function init(items_,boardsUrl_) {
     categoriesData = []
     items.categoryReview.stop()
+
+    var s = items.categories.split(" ")
```
name the variable correctly

petitlapin commented on src/core/FileNames.h:
```diff
@@ -0,0 +1,37 @@
+#ifndef FILENAMES_H
+#define FILENAMES_H
+
+#include <dirent.h>
```
why?

petitlapin commented on src/core/FileNames.cpp:
```diff
@@ -0,0 +1,52 @@
+
+#include "FileNames.h"
+
+#include <dirent.h>
```
why?

petitlapin commented on src/core/FileNames.cpp:
```diff
@@ -0,0 +1,52 @@
+
+#include <dirent.h>
+#include <QString>
+#include <string.h>
```
why?

petitlapin commented on src/core/FileNames.cpp:
```diff
@@ -0,0 +1,52 @@
+{
+}
+
+QString FileNames::getFiles(QString location)
```
use const &

petitlapin commented on src/core/FileNames.cpp:
```diff
@@ -0,0 +1,52 @@
+QString FileNames::getFiles(QString location)
+{
+    const char *s = location.toUtf8().constData();
+    DIR *dpdf = opendir(s);
```
use QDir which is portable

petitlapin commented on src/core/FileNames.cpp:
```diff
@@ -0,0 +1,52 @@
+        while(epdf = readdir(dpdf)) {
+            if(strcmp(epdf->d_name, ".") != 0 && strcmp(epdf->d_name,"..")) {
+                // next file name = epdf->d_name
+                files.append(QString::fromLocal8Bit((epdf->d_name)));
```
use a QList/QStringList

petitlapin commented on src/core/FileNames.cpp:
```diff
@@ -0,0 +1,52 @@
+/* GCompris - FileNames.cpp
```
it would be better to call it Dir or Directory

petitlapin commented on src/core/main.cpp:
```diff
@@ -268,6 +272,9 @@ int main(int argc, char *argv[])
         window->show();
     }
 
+    QScopedPointer<FileNames> files(new FileNames);
```
It would be better to keep the same instanciation as File or other classes

petitlapin commented on src/core/FileNames.h:
```diff
@@ -0,0 +1,37 @@
+#include <QString>
+#include <QObject>
+
+class FileNames : public QObject
```
Add a quick documentation to what the class do (you can check File for example)

divyam3897 commented on src/activities/categorization/resource/board/category_odd_even.qml:
```diff
@@ -23,38 +23,38 @@ import QtQuick 2.0
-                    "maxNumberOfGood": 4,
+                    "image": imagesPrefix + "sixteen.png",
+                    "maxNumberOfGood": 5,
                     "maxNumberOfBad": 4,
```
As mentioned use more images in initial levels (teaching part) . Add 12 images in level1, 9 images in level2, 6 images in level 3 (concept building) , 6 images in level 4 (testing).

divyam3897 commented on src/activities/categorization/resource/board/category_odd_even.qml:
```diff
@@ -0,0 +1,63 @@
+                    "maxNumberOfBad": 5,
+                    "prefix": "qrc:/gcompris/src/activities/categorization/resource/images/odd_even/",
+                    "good": ["even786.svg","even276.svg","even192.svg","even92.svg"],
+                    "bad": ["odd111.svg","odd145.svg","odd39.svg","odd353.svg","odd23.svg"]
```
Did you create the images on your own? If not add where you took it from with the correct licence in README as done for other images.

RudraNilBasu replied:
@divyam3897 yes, all the images in qrc:/gcompris/src/activities/categorization/resource/images/odd_even/ are created by me.

I also used the images from word_sample in the first level.

RudraNilBasu commented:
* Used QDir instead of opendir()
* Removed use of relative paths
* Renamed Filename.h/cpp to Directory.h/cpp
* Added documentation for Directory class
* getFiles(QString) now returns QStringList
* Changed instantiation technique
* No of levels in odd-even category is now 
increased to 6
@petitlapin @divyam3897 can you have a look at it?

petitlapin commented on src/core/main.cpp:
```diff
@@ -268,6 +272,8 @@ int main(int argc, char *argv[])
         window->show();
     }
 
+    engine.rootContext()->setContextProperty("files", new Directory);
```
you don't need this (and the added include on top)

petitlapin commented on src/core/Directory.cpp:
```diff
@@ -0,0 +1,52 @@
+{
+    QStringList fileNames;
+    QDir dir(location);
+    QFileInfoList files=dir.entryInfoList();
```
you can directly use http://doc.qt.io/qt-5/qdir.html#entryList

petitlapin commented on src/core/Directory.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void Directory::init()
+{
+    qmlRegisterType<Directory>("GCompris", 1, 0, "directories");
```
keep the consistency with the other qmlRegisterType

petitlapin commented on src/activities/categorization/categorization.js:
```diff
@@ -51,11 +51,19 @@ function init(items_,boardsUrl_) {
+
+    var categoryLists = []
+    categoryLists = items.categories
+    var allFilesExist = items.file.exists(fileName) ? true : false
```
isEmbeddedMode is better as name

RudraNilBasu commented:
@petitlapin Updated the PR, the following issues should now be fixed
* Removed the use of setContextProperty()
* Used entryList() instead of entryInfoList()
* Renamed allFilesExist -> isEmbeddedMode
* Optimised the size of the images used in
  odd-even category

petitlapin commented:
Thank you for the PR, I commited it in the main repo:
https://cgit.kde.org/gcompris.git/commit/?id=955462b943c34fc130d1a68fcfb0e1ec6393a3f0 (for the directory class)
https://cgit.kde.org/gcompris.git/commit/?id=db7a4a9b743a3521c7c68f0b2b54719cbc9582db (for the level).
I did a few changes, feel free to check them. Most notable is that I put the odd_even category in the "full" version. To test it, you can generate the words.rcc from https://github.com/gcompris/GCompris-words and put it in ~/.cache/KDE/gcompris-qt/data2/words/ and start again GCompris.
